tests/iterator.cpp

```diff
@@ -0,0 +1,60 @@
+#include <algorithm>
+#include <catch2/catch.hpp>
+#include <llama/llama.hpp>
+#include <numeric>
+
+// clang-format off
+namespace tag {
+    struct X {};
+    struct Y {};
+    struct Z {};
+}
+
+using Position = llama::DS<
+    llama::DE<tag::X, int>,
+    llama::DE<tag::Y, int>,
+    llama::DE<tag::Z, int>
+>;
+// clang-format on
+
+TEST_CASE("iterator")
+{
+    using ArrayDomain = llama::ArrayDomain<1>;
+    constexpr auto arrayDomain = ArrayDomain{32};
+    constexpr auto mapping = llama::mapping::AoS<ArrayDomain, Position>{arrayDomain};
+    auto view = llama::allocView(mapping);
+
+    for (auto vd : view)
+    {
+        vd(tag::X{}) = 1;
+        vd(tag::Y{}) = 2;
+        vd(tag::Z{}) = 3;
+    }
+    std::transform(begin(view), end(view), begin(view), [](auto vd) { return vd * 2; });
+    const int sumY = std::accumulate(begin(view), end(view), 0, [](int acc, auto vd) { return acc + vd(tag::Y{}); });
+    CHECK(sumY == 128);
+}
+
+TEST_CASE("iterator.std_copy")
+{
+    using ArrayDomain = llama::ArrayDomain<1>;
+    constexpr auto arrayDomain = ArrayDomain{32};
+    auto aosView = llama::allocView(llama::mapping::AoS<ArrayDomain, Position>{arrayDomain});
+    auto soaView = llama::allocView(llama::mapping::SoA<ArrayDomain, Position>{arrayDomain});
+
+    int i = 0;
+    for (auto vd : aosView)
+    {
+        vd(tag::X{}) = ++i;
+        vd(tag::Y{}) = ++i;
+        vd(tag::Z{}) = ++i;
+    }
+    std::copy(begin(aosView), end(aosView), begin(soaView));
```

Comment:
Having this line work is super crazy! This means we can use `std::copy` to copy between ANY LLAMA layouts. This will always do an element-wise copy, which is sometimes suboptimal. But it correctly shuffles the layouts through the iterators and virtual datums.